docs/Blog-Post.md

```diff
@@ -0,0 +1,276 @@
+    return Attention(proj_qkv, proj_answer)
+
+  def __call__(self, x, mask):
+    qkv_out = self.proj_qkv(x)
+    q, k, v = qkv_out.unbind("qkv")
+
+    # Rename k and v's Pos as Haliax doesn't support unnamed axes or duplicate axes
+    k = k.rename({Pos: KPos})
+    v = v.rename({Pos: KPos})
+
+    answers = attention(Key, KPos, q, k, v, causal_mask)
+
+    x = self.proj_answer(answers)
+    return x
+```
+
+We use named axes both to improve legibility and to enable scale: named axes are the basis of our
+[Fully-Sharded Data Parallel](https://engineering.fb.com/2021/07/15/open-source/fsdp/) implementation. FSDP can
+be added to the training code with about 10 lines of code, enabling scale to at least 256 TPU cores (which is
+as many as we can get our hands on) and at least 65b parameters (which is way bigger than we have compute for).
+
+### Haliax Tutorials
+
+For more details, please see our interactive tutorials on Colab:
+
+* [Introduction to Haliax with Transformers](https://colab.research.google.com/drive/1TiTcQQ4V5mopbgCu1SVl-oqJtXn7rFnC?usp=sharing)
+* [Scaling Transformers in Haliax](https://colab.research.google.com/drive/1QX4yH3zRFF3Xiibf1aahETcSQ5nbcUMz?usp=sharing), including FSDP in Jax.
+
+## Levanter: Bitwise Determinism with Jax
+
+Levanter is a library for training foundation models built on top of Haliax. It provides a complete pipeline
+for training a GPT-2-like Transformer, complete with data preparation, logging, training, checkpointing, evaluation, and export,
+while maintaining bitwise reproducibility throughout.
+
+We have used Levanter to train models as large as 6.7b parameters on a v3-256, and have run experiments showing that it can scale
+up to least 65b parameters.
+
+### Bitwise Reproducibility
+
+One of the benefits of Jax is that it offers strong guarantees for reproducibility. In particular, Jax's fine-grained
+control over random number generation makes it easy to ensure bitwise reproducibility, especially when using TPUs.
+Levanter takes advantage of this to offer bitwise reproducibility for training runs, even after preemption. As an example,
+here is a screenshot of a training run being resumed multiple times, even on different TPU pod slices:
+
+![plot showing bitwise reproducibility with four training runs superimposed with the exact same loss curve](figures/bitwise_repro_curve.png)
+
+The fact that you can't make out the different lines is the point: the training runs are bitwise identical,
+a huge advantage for debugging and reproducibility.
+
+### Efficiency and Scale
+
+XXX something something v3-256 scaling numbers?
+
+### Other Features
+
+* **Preprocessing**: Levanter uses [Hugging Face Tokenizers](https://github.com/huggingface/tokenizers/) to preprocess text
+using distributed preprocessing backed by [Ray](https://www.ray.io/))
+* **Training**: In addition to Jax and Haliax, Levanter uses [Optax](https://github.com/deepmind/optax) for optimization,
```

Comment:
Can Sophia not be released as part of Levanter?

Reply:
Sophia-G isn't quite working in Jax yet. Still figuring it out.

Reply:
(turns out it was just a bad gamma number. Hong is training this)